compiler/src/test/scala/org/scalajs/core/compiler/test/JSOptionalTest.scala

```diff
@@ -0,0 +1,160 @@
+      override val a2: js.UndefOr[Int] = js.undefined
+
+      override def b1: js.UndefOr[Int] = js.undefined
+      override def b2: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    s"""
+      |newSource1.scala:16: error: Cannot override concrete val a1: scala.scalajs.js.UndefOr[Int] with optional override val a1: scala.scalajs.js.UndefOr[Int] (an optional member is one defined in a non-native JS trait with `= js.undefined`).
+      |      override val a1: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:19: error: Cannot override concrete def b1: scala.scalajs.js.UndefOr[Int] with optional override def b1: scala.scalajs.js.UndefOr[Int] (an optional member is one defined in a non-native JS trait with `= js.undefined`).
+      |      override def b1: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
+
+    s"""
+    @js.native
+    class A extends js.Object {
+      val a: js.UndefOr[Int] = js.native
+      def b: js.UndefOr[Int] = js.native
+    }
+
+    @ScalaJSDefined
+    trait B extends A {
+      override val a: js.UndefOr[Int] = js.undefined
+      override def b: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    s"""
+      |newSource1.scala:13: error: Cannot override concrete val a: scala.scalajs.js.UndefOr[Int] with optional override val a: scala.scalajs.js.UndefOr[Int] (an optional member is one defined in a non-native JS trait with `= js.undefined`).
+      |      override val a: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:14: error: Cannot override concrete def b: scala.scalajs.js.UndefOr[Int] with optional override def b: scala.scalajs.js.UndefOr[Int] (an optional member is one defined in a non-native JS trait with `= js.undefined`).
+      |      override def b: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
+
+    s"""
+    @js.native
+    trait A extends js.Object {
+      val a: js.UndefOr[Int] = js.native
+      def b: js.UndefOr[Int] = js.native
+    }
+
+    @js.native
+    class B extends A
+
+    @ScalaJSDefined
+    trait C extends B {
+      override val a: js.UndefOr[Int] = js.undefined
+      override def b: js.UndefOr[Int] = js.undefined
+    }
+    """ hasErrors
+    s"""
+      |newSource1.scala:16: error: Cannot override concrete val a: scala.scalajs.js.UndefOr[Int] with optional override val a: scala.scalajs.js.UndefOr[Int] (an optional member is one defined in a non-native JS trait with `= js.undefined`).
+      |      override val a: js.UndefOr[Int] = js.undefined
+      |                   ^
+      |newSource1.scala:17: error: Cannot override concrete def b: scala.scalajs.js.UndefOr[Int] with optional override def b: scala.scalajs.js.UndefOr[Int] (an optional member is one defined in a non-native JS trait with `= js.undefined`).
+      |      override def b: js.UndefOr[Int] = js.undefined
+      |                   ^
+    """
```

Comment:
That's prohibited altogether. A Scala.js-defined thing can never directly extend a native JS trait.